core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,248 @@
+    // API spec - https://github.com/ethereum/keymanager-APIs/tree/master/flows#import
+    final RequestParameters params = context.get("parsedParameters");
+    final ImportKeystoresRequestBody parsedBody;
+    try {
+      parsedBody = parseRequestBody(params);
+    } catch (final IllegalArgumentException | JsonProcessingException e) {
+      handleInvalidRequest(context, e);
+      return;
+    }
+
+    // check that keystores have matching passwords
+    if (parsedBody.getKeystores().size() != parsedBody.getPasswords().size()) {
+      context.fail(BAD_REQUEST);
+      return;
+    }
+
+    // no keystores passed in, nothing to do, return 200 and empty response.
+    if (parsedBody.getKeystores().isEmpty()) {
+      try {
+        context
+            .response()
+            .putHeader(CONTENT_TYPE, JSON_UTF_8)
+            .setStatusCode(SUCCESS)
+            .end(
+                objectMapper.writeValueAsString(
+                    new ImportKeystoresResponse(Collections.emptyList())));
+      } catch (JsonProcessingException e) {
+        context.fail(SERVER_ERROR, e);
+      }
+      return;
+    }
+
+    // extract pubkeys to import first
+    final List<String> pubkeysToImport = new ArrayList<>();
+    try {
+      pubkeysToImport.addAll(
+          parsedBody.getKeystores().stream()
+              .map(json -> new JsonObject(json).getString("pubkey"))
+              .map(
+                  key ->
+                      key.startsWith("0x")
+                          ? key
+                          : "0x" + key) // always use 0x prefix for comparisons
+              .collect(Collectors.toList()));
+    } catch (Exception e) {
+      context.fail(BAD_REQUEST, e);
+      return;
+    }
+
+    // read slashing protection data if present and import data matching keys to import only
+    if (slashingProtection.isPresent()
+        && !StringUtils.isEmpty(parsedBody.getSlashingProtection())) {
+      try {
+        final InputStream slashingProtectionData =
+            new ByteArrayInputStream(
+                parsedBody.getSlashingProtection().getBytes(StandardCharsets.UTF_8));
+        slashingProtection.get().importDataWithFilter(slashingProtectionData, pubkeysToImport);
+      } catch (Exception e) {
+        context.fail(BAD_REQUEST, e);
+        return;
+      }
+    }
+
+    final Set<String> existingPubkeys = artifactSignerProvider.availableIdentifiers();
+    final List<ImportKeystoreResult> results = new ArrayList<>();
+    for (int i = 0; i < parsedBody.getKeystores().size(); i++) {
+      final String pubkey = pubkeysToImport.get(i);
+      try {
+        final String jsonKeystoreData = parsedBody.getKeystores().get(i);
+        final String password = parsedBody.getPasswords().get(i);
+
+        if (existingPubkeys.contains(pubkey)) {
+          // keystore already loaded
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.DUPLICATE, null));
+        } else {
+          // new keystore to import
```

Comment:
Think need a test that shows that successful keys are imported even when there are failures and some duplicates and that the keys for failing imports are removed.

Reply:
done